### PR TITLE
Add gpt-oss

### PR DESCRIPTION
**Pull request by cuichenx**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add gpt-oss model
Use with this megatron branch:
https://github.com/NVIDIA/Megatron-LM/tree/chcui/gpt_oss

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/model/base.py [hunk 1]
```diff
@@ -384,6 +386,10 @@ def configure_model(self, tokenizer, pre_process=None, post_process=None, vp_sta
             kwargs = {"mtp_block_spec": mtp_block_spec(self, vp_stage=vp_stage)}
         else:
             kwargs = {}
+
+        if self.attention_backend == AttnBackend.local and getattr(self, "softmax_type", "vanilla") != "vanilla":
```
Comment (guyueh1) [inline nemo/collections/llm/gpt/model/base.py]: I think as long as attention_backend == local, we should change the DPA module to MCORE DPA, regardless of softmax_type. Is it necessary to put softmax_type != "vanilla" here?

Comment (cuichenx) [reply nemo/collections/llm/gpt/model/base.py]: Good catch! I'll remove the condition

File: scripts/llm/generate.py [hunk 2]
```diff
@@ -165,7 +169,7 @@ def get_args():
         tensor_model_parallel_size=args.tp,
         pipeline_model_parallel_size=args.pp,
         expert_model_parallel_size=args.ep,
-        expert_tensor_parallel_size=1 if args.ep > 1 else None,
```
Comment (suiyoubi) [inline scripts/llm/generate.py:168]: why do we need to modify this ? Does ETP not work for the inference ? By default I think it uses the same as the TP when etp=None

File: nemo/lightning/io/state.py [hunk 3]
```diff
@@ -222,11 +224,17 @@ def scale_weights(ctx):
         f"Did you forget to include these parameters in the mapping or transforms in `convert_state`?"
     )
 
-    assert target_orig_dtypes == extract_dtypes(_target.named_parameters()), (
-        f"dtype mismatch between source and target state dicts. "
-        f"Left side is { {k: v for k, v in target_orig_dtypes.items() if v!=torch.bfloat16} }, "
-        f"Right side is { {k: v for k, v in extract_dtypes(_target.named_parameters()).items() if v!=torch.bfloat16} }"
-    )
+    if cast_dtype:
```
Comment (suiyoubi) [inline nemo/lightning/io/state.py:227]: I thought this logic is merged in some previous PR no ?

File: nemo/collections/llm/recipes/gpt_oss_20b.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+        Python API usage:
+            >>> recipe = finetune_recipe(name="gpt_oss_20b_finetune", num_nodes=1)
+            >>> print(recipe)
+
+    Note:
+        This recipe uses the SQuAD dataset for fine-tuning.
+    """
+    recipe = default_finetune_recipe(model(), resume_path, dir, name, num_nodes, num_gpus_per_node, packed_sequence)
+    # recipe.data.dataset_kwargs = {'add_bos': True, "get_attention_mask_from_fusion": True}
```
Comment (suiyoubi) [inline nemo/collections/llm/recipes/gpt_oss_20b.py]: Remove the commented out code

File: nemo/collections/llm/recipes/gpt_oss_20b.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    peft_scheme: Optional[str] = 'lora',
+    packed_sequence: bool = False,
+) -> run.Partial:
+    """
+    Create a fine-tuning recipe for GPT-OSS 20B model.
+
+    This function sets up a complete configuration for fine-tuning, including
+    model, trainer, data, logging, optimization, and resumption settings.
+    The recipe uses LoRA (Low-Rank Adaptation) for efficient fine-tuning, unless peft_scheme is set to None.
+    This model uses Mixture of Experts (MoE) architecture with 128 experts.
```
Comment (suiyoubi) [inline nemo/collections/llm/recipes/gpt_oss_20b.py]: 32 experts

File: nemo/collections/llm/peft/api.py [hunk 6]
```diff
@@ -124,6 +124,7 @@ def _load_base_model_and_lora(lora_checkpoint_path: Path) -> Tuple[pl.LightningM
     model.model_transform, model.__io__.model_transform = None, None
     model.config.bf16 = True
     model.config.params_dtype = torch.bfloat16
+    model.config.perform_initialization = False
```
Comment (suiyoubi) [inline nemo/collections/llm/peft/api.py:127]: Could you explain a bit what does this line do ?

File: nemo/collections/llm/gpt/model/gpt_oss.py [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+    moe_ffn_hidden_size: int = 2880
+    moe_router_load_balancing_type: str = "none"
+    seq_length: int = 131072
+    window_size: Optional[Tuple[int, int]] = (128, 0)
+    softmax_type: Literal['vanilla', 'off-by-one', 'learnable'] = "learnable"
+    activation_func: Callable = quick_gelu
+    glu_linear_offset: float = 1.0
+    bias_activation_fusion: bool = True
+    window_attn_skip_freq: Optional[Union[int, List[int]]] = 2  # alternative SWA/full
+    attention_backend: AttnBackend = AttnBackend.local  # supports "local" and "fused"
```
Comment (suiyoubi) [inline nemo/collections/llm/gpt/model/gpt_oss.py]: if local and fused are both supported, why don't we use fused for better perf ?

File: nemo/collections/llm/gpt/model/gpt_oss.py [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+@io.model_importer(GPTOSSModel, "hf")
+class HFGPTOSSImporter(_BaseGPTOSSImporter):
+    """Importer for GPT-OSS models from Hugging Face"""
+
+    # pylint: disable=C0115,C0116
+    def apply(self, output_path: Path) -> Path:
+        logging.setLevel(logging.DEBUG)
+        source_state = self.hf_ckpt_load()
+        source = _ModelState(source_state)
+        # source = AutoModelForCausalLM.from_pretrained(str(self), torch_dtype='auto', trust_remote_code=True)
```
Comment (suiyoubi) [inline nemo/collections/llm/gpt/model/gpt_oss.py]: remove commented out line

File: nemo/collections/llm/gpt/model/gpt_oss.py [hunk 9]
```diff
@@ -0,0 +1,522 @@
...
+                    "**.self_attn.k_proj.weight",
+                    "**.self_attn.v_proj.weight",
+                ),
+                target_key="**.self_attention.linear_qkv.weight",
+                fn=TransformFns.merge_qkv,
+            ),
+            io.state_transform(
+                source_key="**.mlp.experts.gate_up_proj_bias",
+                target_key="**.mlp.experts.linear_fc1.bias*",
+                fn=lambda x: [_interleave(e) for e in [*x]],
```
Comment (suiyoubi) [inline nemo/collections/llm/gpt/model/gpt_oss.py:282]: This is cool :D

File: nemo/collections/llm/gpt/model/gpt_oss.py [hunk 10]
```diff
@@ -0,0 +1,522 @@
...
+        target = self.init(torch_dtype_from_mcore_config(source.config))
+        target = self.convert_state(source, target)
+
+        target = target.cpu()
+        target.save_pretrained(output_path)
+        self.tokenizer.save_pretrained(output_path)
+
+        return output_path
+
+    def convert_state(self, source, target):
```
Comment (suiyoubi) [inline nemo/collections/llm/gpt/model/gpt_oss.py:440]: Just wondering, for model import you need to de-quantize some weights, and we don't need to do the opposite for export ?

**Review by guyueh1 — DISMISSED** [review]

File: scripts/llm/generate.py (hunk 2 above)
Comment (cuichenx) [reply scripts/llm/generate.py:168]: ETP doesn't work for this model (with expert bias), so I disabled it here. I can modify it to disable only when expert bias is true.

File: nemo/lightning/io/state.py (hunk 3 above)
Comment (cuichenx) [reply nemo/lightning/io/state.py:227]: yes :sweat_smile: https://github.com/NVIDIA/NeMo/pull/14307
Once that's merged this change should disappear.

File: nemo/collections/llm/peft/api.py (hunk 6 above)
Comment (cuichenx) [reply nemo/collections/llm/peft/api.py:127]: Because we're loading model weights, we don't need to initialize weights (the initialized weights will be overwritten by loaded weights). In practice adding this removes some CPU RNG state warnings and cuts down lora merging time from 7min to 1min.

File: nemo/collections/llm/gpt/model/gpt_oss.py (hunk 7 above)
Comment (cuichenx) [reply nemo/collections/llm/gpt/model/gpt_oss.py]: Thanks for catching this, actually only local is supported as of this PR. I'll change it.

File: nemo/collections/llm/gpt/model/gpt_oss.py (hunk 10 above)
Comment (cuichenx) [reply nemo/collections/llm/gpt/model/gpt_oss.py:440]: good question, for now we'll only support BF16 HF weights. We'll add support for quantized export soon.

File: scripts/llm/generate.py (hunk 2 above)
Comment (cuichenx) [reply scripts/llm/generate.py:168]: ETP doesn't work with this model (with expert bias). But you're right I shouldn't disable it for all models. I will modify the condition
